cli/js/repl.ts

```diff
@@ -106,6 +107,13 @@ export async function replLoop(): Promise<void> {
     },
   });
 
+  // Configure globalThis.close to terminate REPL properly when `close()` is called.
+  Object.defineProperty(
+    globalThis,
+    "close",
+    writable(() => quitRepl(0))
+  );
```

Comment:
I also noticed this problem - it's annoying - so thanks for looking into it. But I'm not sure I agree with you fix. It works, of course, but I wonder if there isn't a better solution. I don't like that `close()` in the REPL is now a different function than it is in scripts.

Reply:
@ry Ok, I agree with you.

How about the following changes?

- Stop overwriting `globalThis.close()`.
- Add a function that checks if `close()` is called (`isCloseCalled()`).
- Quit the REPL if `isCloseCalled()` returns `true` before entering the next loop .

```diff
diff --git a/cli/js/repl.ts b/cli/js/repl.ts
index d969a1e3..b6b83a02 100644
--- a/cli/js/repl.ts
+++ b/cli/js/repl.ts
@@ -33,6 +33,12 @@ function isRecoverableError(e: Error): boolean {
   return recoverableErrorMessages.includes(e.message);
 }

+// Returns `true` if `close()` is called in REPL.
+function isCloseCalled(): boolean {
+  // @ts-ignore
+  return globalThis.closed;
+}
+
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 type Value = any;

@@ -46,7 +52,9 @@ function evaluate(code: string): boolean {
   const [result, errInfo] = core.evalContext(code);
   if (!errInfo) {
     lastEvalResult = result;
-    replLog(result);
+    if (!isCloseCalled()) {
+      replLog(result);
+    }
   } else if (errInfo.isCompileError && isRecoverableError(errInfo.thrown)) {
     // Recoverable compiler error
     return false; // don't consume code.
@@ -107,17 +115,14 @@ export async function replLoop(): Promise<void> {
     },
   });

-  // Configure globalThis.close to terminate REPL properly when `close()` is called.
-  Object.defineProperty(
-    globalThis,
-    "close",
-    writable(() => quitRepl(0))
-  );
-
   replLog(`Deno ${version.deno}`);
   replLog("exit using ctrl+d or close()");

   while (true) {
+    if (isCloseCalled()) {
+      quitRepl(0);
+    }
+
     let code = "";
     // Top level read
     try {

```

Reply:
Note: `globalThis.closed` is defined in [runtime_main.ts](https://github.com/denoland/deno/blob/89fe81168e75a375a67e9d391b367d2a9749153c/cli/js/runtime_main.ts#L64).